### PR TITLE
#65: Group values by weeks in QoS graph

### LINKED ISSUE #65: group values by weeks, in the QoS graph
Currently, in the `qos.xsl`, we render all available facts, no matter what are the dates of them. Instead, let's group by weeks and only show one fact per week. Simply put, if we have two facts with the `when` property inside the same week, we show only one of them, ignoring the second one.

### PR DESCRIPTION
**Pull request by fruxxxl**
## Summary

Fixes #65

Currently in `qo-section.xsl`, all available facts are rendered regardless of their dates. This PR groups facts by ISO week using `xsl:for-each-group` with `format-dateTime(..., '[Y0001]-[W01]')` and keeps only the last fact per week.

## Changes

- Added `$raw` variable for initial fact filtering
- Added `$facts` variable with `xsl:for-each-group` grouping by ISO year-week
- From each group, only the last fact (`current-group()[last()]`) is kept
- Updated all references from `$facts` to `$facts/f` to work with the new variable structure

## How to test

1. Run unit tests: `bundle exec rake test` — all pass
2. Build assets: `make assets`
3. Generate HTML: `make target/html/single-qos.html`
4. With test data containing multiple facts in the same week, only one point per week should appear on the graph

## Files changed

- `xsl/qo-section.xsl` — weekly grouping of facts

## Summary by CodeRabbit

* **New Features**
  * Chart labels and weekly summaries now use ISO-week labels (YYYY-Www).

* **Refactor**
  * Data is downsampled to one representative entry per ISO week; labels and datasets are generated from these weekly groups.

* **Bug Fixes**
  * Correct handling of ISO week/year boundaries and deterministic within-week ordering for representative selection.

* **Tests**
  * Added unit tests and a weekly fixture covering ISO-week edge cases and ordering.

### CODE REVIEW

**Comment by fruxxxl** [comment]
@coderabbitai re-review

**Comment by fruxxxl** [comment]
@coderabbitai re-review

**Comment by fruxxxl** [comment]
Addressed both review findings in commit 3351980:

**1. ISO calendar for week grouping**

The suggested fix was to use `format-dateTime(xs:dateTime(when), '[Y0001]-[W01]', (), 'ISO')` — the 4th argument `calendar='ISO'`. However, this project uses **Saxon-HE 9.8** which is an **XSLT 2.0** processor. The 4-argument form of `format-dateTime` requires **XPath 3.0 / XSLT 3.0** (Saxon-HE 10+). Passing 4 arguments causes a compile-time error:

```
XPST0017: Cannot find a 4-argument function named
{http://www.w3.org/2005/xpath-functions}format-dateTime()
```

Instead, I added a custom `z:iso-week()` function that computes the correct ISO year-week using the **Thursday algorithm**:

1. Compute day-of-week (Monday=1 … Sunday=7) from a known Monday reference date (`2024-01-01`)
2. Find the **Thursday of the same ISO week** as the input date — this is the key trick: the calendar year of this Thursday is **guaranteed to equal the ISO week-numbering year** (by definition, since ISO weeks with Thursday in year N belong to year N)
3. Combine `year-from-date($thu)` with `format-date($thu, '[W01]')` for the week number

Verified boundary cases: `2023-12-31` → `2024-W01`, `2024-12-31` → `2025-W01`, `2024-01-01` → `2024-W01`.

**2. Chronological selection within groups**

Replaced `current-group()[last()]` (document order) with `xsl:sort by xs:dateTime(when) ascending` + `position() = last()` to pick the chronologically latest fact per week.

**Comment by fruxxxl** [comment]
@coderabbitai re-review

File: xsl/qo-section.xsl [hunk 1]
```diff
@@ -15,12 +15,19 @@
...
     <xsl:param name="before"/>
-    <xsl:variable name="facts" select="/fb/f[what=$what and xs:dateTime(when) &gt; (xs:dateTime($today) - xs:dayTimeDuration('P180D'))]"/>
+    <xsl:variable name="raw" select="/fb/f[what=$what and xs:dateTime(when) &gt; (xs:dateTime($today) - xs:dayTimeDuration('P180D'))]"/>
+    <xsl:variable name="facts">
+      <xsl:for-each-group select="$raw" group-by="format-dateTime(xs:dateTime(when), '[Y0001]-[W01]')">
+        <xsl:for-each select="current-group()[last()]">
+          <xsl:copy-of select="."/>
+        </xsl:for-each>
+      </xsl:for-each-group>
+    </xsl:variable>
```
[the remark this answers is not in the document]
Comment (fruxxxl) [reply xsl/qo-section.xsl:38]: Fixed in squashed commit fb6a027. All issues addressed: ISO week function with signed Thursday offset, timezone-aware reference date, chronological sort for group representative, correct test expectations.

File: test/pages/test_qo_section.rb [hunk 2]
```diff
@@ -25,4 +25,36 @@ def test_snake_case_to_title
...
+    assert_equal('2024-W29', xml.xpath('//r/text()').to_s.strip)
+  end
+
+  def test_iso_week_year_boundary_dec31
+    xml = xslt(
+      '<r><xsl:value-of select="z:iso-week(xs:dateTime(\'2023-12-31T00:00:00Z\'))"/></r>',
+      '<fb/>'
+    )
+    assert_equal('2024-W01', xml.xpath('//r/text()').to_s.strip)
+  end
```
[the remark this answers is not in the document]
Comment (fruxxxl) [reply test/pages/test_qo_section.rb:43]: Fixed in squashed commit fb6a027. All issues addressed: ISO week function with signed Thursday offset, timezone-aware reference date, chronological sort for group representative, correct test expectations.

File: tests/weekly.yml [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+  where: github
+  when: 2024-06-26T22:22:22Z
+-
+  # Week 2024-W01: 3 facts in SAME week, document order REVERSED (newest first)
+  # Without fix: would pick _id:10 (composit=100) as "last" by doc order
+  # With fix: picks _id:12 (composite=200) as chronologically latest
+  _id: 10
+  when: 2024-01-03T10:00:00Z
+  what: quality-of-service
+  composite: 100
```
[the remark this answers is not in the document]
Comment (fruxxxl) [reply tests/weekly.yml]: Fixed in squashed commit fb6a027. All issues addressed: ISO week function with signed Thursday offset, timezone-aware reference date, chronological sort for group representative, correct test expectations.

File: tests/weekly.yml [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+-
+  # Year boundary: Dec 31 2023 = ISO 2024-W01
+  # Without fix: groups as "2023-W01" (wrong!), separate from Jan 1-3 facts
+  # With fix: groups as "2024-W01" together with Jan 1-3 facts
+  _id: 15
+  when: 2023-12-31T22:00:00Z
+  what: quality-of-service
+  composite: 70
+  average_issue_lifetime: 60
+  average_pull_lifetime: 150
```
[the remark this answers is not in the document]
Comment (fruxxxl) [reply tests/weekly.yml]: Fixed in squashed commit fb6a027. All issues addressed: ISO week function with signed Thursday offset, timezone-aware reference date, chronological sort for group representative, correct test expectations.

File: xsl/qo-section.xsl [hunk 5]
```diff
@@ -5,6 +5,15 @@
...
 <xsl:stylesheet xmlns:xsl="http://www.w3.org/1999/XSL/Transform" version="2.0" xmlns:xs="http://www.w3.org/2001/XMLSchema" xmlns:z="https://www.zerocracy.com" exclude-result-prefixes="xs z">
   <xsl:include href="script-with-cdata.xsl"/>
+  <xsl:function name="z:iso-week" as="xs:string">
+    <xsl:param name="dt" as="xs:dateTime"/>
+    <xsl:variable name="d" select="xs:date($dt)"/>
+    <xsl:variable name="days" select="xs:integer(($d - xs:date('2024-01-01Z')) div xs:dayTimeDuration('P1D'))"/>
+    <xsl:variable name="dow" select="($days mod 7 + 7) mod 7 + 1"/>
+    <xsl:variable name="off" select="4 - $dow"/>
+    <xsl:variable name="thu" select="if ($off ge 0) then $d + xs:dayTimeDuration(concat('P', $off, 'D')) else $d - xs:dayTimeDuration(concat('P', -$off, 'D'))"/>
+    <xsl:value-of select="concat(format-integer(year-from-date($thu), '0001'), '-W', format-date($thu, '[W01]'))"/>
```
[the remark this answers is not in the document]
Comment (fruxxxl) [reply xsl/qo-section.xsl]: Fixed in squashed commit fb6a027. All issues addressed: ISO week function with signed Thursday offset, timezone-aware reference date, chronological sort for group representative, correct test expectations.

File: xsl/qo-section.xsl [hunk 6]
```diff
@@ -5,6 +5,15 @@
 -->
 <xsl:stylesheet xmlns:xsl="http://www.w3.org/1999/XSL/Transform" version="2.0" xmlns:xs="http://www.w3.org/2001/XMLSchema" xmlns:z="https://www.zerocracy.com" exclude-result-prefixes="xs z">
   <xsl:include href="script-with-cdata.xsl"/>
+  <xsl:function name="z:iso-week" as="xs:string">
+    <xsl:param name="dt" as="xs:dateTime"/>
+    <xsl:variable name="d" select="xs:date($dt)"/>
+    <xsl:variable name="days" select="xs:integer(($d - xs:date('2024-01-01Z')) div xs:dayTimeDuration('P1D'))"/>
```
[the remark this answers is not in the document]
Comment (fruxxxl) [reply xsl/qo-section.xsl]: Fixed in squashed commit fb6a027.

File: xsl/qo-section.xsl [hunk 7]
```diff
@@ -5,6 +5,14 @@
...
 <xsl:stylesheet xmlns:xsl="http://www.w3.org/1999/XSL/Transform" version="2.0" xmlns:xs="http://www.w3.org/2001/XMLSchema" xmlns:z="https://www.zerocracy.com" exclude-result-prefixes="xs z">
   <xsl:include href="script-with-cdata.xsl"/>
+  <xsl:function name="z:iso-week" as="xs:string">
+    <xsl:param name="dt" as="xs:dateTime"/>
+    <xsl:variable name="d" select="xs:date($dt)"/>
+    <xsl:variable name="days" select="xs:integer(($d - xs:date('2024-01-01')) div xs:dayTimeDuration('P1D'))"/>
+    <xsl:variable name="dow" select="($days mod 7 + 7) mod 7 + 1"/>
+    <xsl:variable name="thu" select="$d + xs:dayTimeDuration(concat('P', (4 - $dow + 7) mod 7, 'D'))"/>
+    <xsl:value-of select="concat(format-integer(year-from-date($thu), '0001'), '-W', format-date($thu, '[W01]'))"/>
+  </xsl:function>
```
[the remark this answers is not in the document]
Comment (fruxxxl) [reply xsl/qo-section.xsl:17]: Fixed in squashed commit fb6a027.

**Comment by fruxxxl** [comment]
@coderabbitai re-review

**Comment by fruxxxl** [comment]
@coderabbitai review

**Comment by fruxxxl** [comment]
@coderabbitai review

**Comment by fruxxxl** [comment]
@yegor256 Take a look please

**Comment by yegor256** [comment]
@rultor merge
